src/ng/directive/ngRepeat.js

```diff
@@ -425,126 +419,124 @@ var ngRepeatDirective = ['$parse', '$animate', '$compile', function($parse, $ani
 
         //watch props
         $scope.$watchCollection(rhs, function ngRepeatAction(collection) {
-          var index, length,
-              previousNode = $element[0],     // node that cloned nodes should be inserted after
-                                              // initialized to the comment node anchor
-              nextNode,
-              // Same as lastBlockMap but it has the current state. It will become the
-              // lastBlockMap on the next iteration.
-              nextBlockMap = createMap(),
-              collectionLength,
-              key, value, // key/value of iteration
-              trackById,
-              trackByIdFn,
-              collectionKeys,
-              block,       // last object information {scope, element, id}
-              nextBlockOrder,
-              elementsToRemove;
+          var
+            block,                       // last object information {scope, element, id}
+            collectionIsLikeArray,
+            collectionKeys = [],
+            elementsToRemove,
+            index, key, value,
+            lastBlockOrder = [],
+            lastKey,
+            nextBlockMap = createMap(),
+            nextBlockOrder = [],
+            nextKey,
+            nextLength,
+            previousNode = $element[0],  // node that cloned nodes should be inserted after
+                                         // initialized to the comment node anchor
+            trackById,
+            trackByIdFn;
 
           if (aliasAs) {
             $scope[aliasAs] = collection;
           }
 
-          if (isArrayLike(collection)) {
-            collectionKeys = collection;
+          // get collectionKeys
+          collectionIsLikeArray = isArrayLike(collection);
+          if (collectionIsLikeArray) {
             trackByIdFn = trackByIdExpFn || trackByIdArrayFn;
           } else {
             trackByIdFn = trackByIdExpFn || trackByIdObjFn;
             // if object, extract keys, in enumeration order, unsorted
-            collectionKeys = [];
-            for (var itemKey in collection) {
-              if (hasOwnProperty.call(collection, itemKey) && itemKey.charAt(0) !== '$') {
-                collectionKeys.push(itemKey);
+            for (key in collection) {
+              if (hasOwnProperty.call(collection, key) && key.charAt(0) !== '$') {
+                collectionKeys.push(key);
               }
             }
           }
+          nextLength = collectionIsLikeArray ? collection.length : collectionKeys.length;
 
-          collectionLength = collectionKeys.length;
-          nextBlockOrder = new Array(collectionLength);
-
-          // locate existing items
-          for (index = 0; index < collectionLength; index++) {
-            key = (collection === collectionKeys) ? index : collectionKeys[index];
+          // setup nextBlockMap
```

Comment:
This loop has a similar purpose to the "locate existing items" loop in the previous version.